packages/@jsii/runtime/lib/sync-stdio.ts

```diff
@@ -18,45 +18,41 @@ export class SyncStdio {
   }
 
   public readLine(): string | undefined {
-    if (this.inputQueue.length > 0) {
-      return this.inputQueue.shift();
-    }
-
-    try {
-      const buff = Buffer.alloc(INPUT_BUFFER_SIZE);
-      const read = fs.readSync(STDIN_FD, buff, 0, buff.length, null);
+    const buff = Buffer.alloc(INPUT_BUFFER_SIZE);
+    while (this.inputQueue.length === 0) {
+      try {
+        const read = fs.readSync(STDIN_FD, buff, 0, buff.length, null);
 
-      if (read === 0) {
-        return undefined;
-      }
+        if (read === 0) {
+          return undefined;
+        }
 
-      const str = buff.slice(0, read).toString();
+        const str = buff.slice(0, read).toString();
 
-      for (const ch of str) {
-        if (ch === '\n') {
-          this.inputQueue.push(this.currentLine);
-          this.currentLine = '';
-        } else {
-          this.currentLine += ch;
+        for (const ch of str) {
+          if (ch === '\n') {
+            this.inputQueue.push(this.currentLine);
+            this.currentLine = '';
+          } else {
+            this.currentLine += ch;
+          }
+        }
+      } catch (e) {
+        // HACK: node may set O_NONBLOCK on it's STDIN depending on what kind of input it is made
+        // of (see https://github.com/nodejs/help/issues/2663). When STDIN has O_NONBLOCK, calls may
+        // result in EAGAIN. In such cases, the call should be retried until it succeeds. This kind
+        // of polling will result in horrible CPU thrashing, but there does not seem to be a way to
+        // force a O_SYNC access to STDIN in a reliable way within node.
+        // In order to make this stop we need to either stop depending on synchronous reads, or to
```

Comment:
> stop depending on synchronous reads

That would effectively mean callbacks would stop working, right?

Reply:
Why no. This means asynchronous user-code suddenly would throw everything off. Pure-async kernel would work okay if we could guarantee client code NEVER does anything asynchronous by itself.